File: livevideostreamproducer/src/main/java/constantin/livevideostreamproducer/MainActivity.java
package constantin.livevideostreamproducer;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.preference.PreferenceManager;

import android.Manifest;
import android.annotation.SuppressLint;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.EditText;
import android.widget.Toast;

import com.mapzen.prefsplusx.DefaultPreferenceFragment;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

//Request all permissions
//start AVideoStream activity if button is clicked

public class MainActivity extends AppCompatActivity {
    private static final String[] REQUIRED_PERMISSION_LIST = new String[]{
            Manifest.permission.WRITE_EXTERNAL_STORAGE,
            Manifest.permission.ACCESS_FINE_LOCATION,
            Manifest.permission.CAMERA,
    };
    private final List<String> missingPermission = new ArrayList<>();
    private static final int REQUEST_PERMISSION_CODE = 12345;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);


        checkAndRequestPermissions();
        final Context context=this;
        findViewById(R.id.start_stream).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent=new Intent().setClass(context,AVideoStream.class);
                startActivity(intent);
            }
        });
        final EditText editText=findViewById(R.id.ip_address_edit_text);
        editText.setText(getIpAddress(context));
        editText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
            }

            @SuppressLint("ApplySharedPref")
            @Override
            public void afterTextChanged(Editable s) {
                writeIpAddress(context,s.toString());
                Toast.makeText(context,"Set IP "+s.toString(),Toast.LENGTH_SHORT).show();
            }
        });
        findViewById(R.id.autofill_button).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                final List<String> ips=IPResolver.findAllPossibleClientIps(context);
                if(ips.size()!=0){
                    final CharSequence[] items =new CharSequence[ips.size()];
                    for(int i=0;i<ips.size();i++){
                        items[i]=ips.get(i)+"";
                    }
                    AlertDialog.Builder builder = new AlertDialog.Builder(MainActivity.this);
                    builder.setTitle("Select destination IP");
                    builder.setItems(items, new DialogInterface.OnClickListener() {
                        public void onClick(DialogInterface dialog, int item) {
                            final String ip=ips.get(item);
                            editText.setText(ip);
                            writeIpAddress(context,ip);
                            Toast.makeText(context,"Set ip to "+ip,Toast.LENGTH_LONG).show();
                        }
                    });
                    AlertDialog alert = builder.create();
                    alert.show();
                }else{
                    Toast.makeText(context,"Cannot autofill ip",Toast.LENGTH_LONG).show();
                }
            }
        });
    }

    @SuppressLint("ApplySharedPref")
    private static void writeIpAddress(final Context context, final String ip){
        PreferenceManager.getDefaultSharedPreferences(context).edit().
                putString(context.getString(R.string.KEY_SP_UDP_IP),ip).commit();
    }
    @SuppressLint("ApplySharedPref")
    private static String getIpAddress(final Context context){
        return PreferenceManager.getDefaultSharedPreferences(context).
                getString(context.getString(R.string.KEY_SP_UDP_IP),"192.168.1.1");
    }

    //Permissions stuff
    private void checkAndRequestPermissions(){
        missingPermission.clear();
        for (String eachPermission : REQUIRED_PERMISSION_LIST) {
            if (ContextCompat.checkSelfPermission(this, eachPermission) != PackageManager.PERMISSION_GRANTED) {
                missingPermission.add(eachPermission);
            }
        }
        if (!missingPermission.isEmpty()) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                final String[] asArray=missingPermission.toArray(new String[0]);
                Log.d("PermissionManager","Request: "+ Arrays.toString(asArray));
                ActivityCompat.requestPermissions(this, asArray, REQUEST_PERMISSION_CODE);
            }
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode,
                                           @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);

        // Check for granted permission and remove from missing list
        if (requestCode == REQUEST_PERMISSION_CODE) {
            for (int i = grantResults.length - 1; i >= 0; i--) {
                if (grantResults[i] == PackageManager.PERMISSION_GRANTED) {
                    missingPermission.remove(permissions[i]);
                }
            }
        }
        if (!missingPermission.isEmpty()) {
            checkAndRequestPermissions();
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.settings, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_settings:
                // User chose the "Settings" item, show the app settings UI...
                Intent i=new Intent();
                i.setClass(this,ASettings.class);
                startActivity(i);
                return true;
            default:
                // If we got here, the user's action was not recognized.
                // Invoke the superclass to handle it.
                return super.onOptionsItemSelected(item);

        }
    }

}

File: Example/src/main/java/constantin/video/example/MainActivity.kt
package constantin.video.example

import android.Manifest
import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.ArrayAdapter
import androidx.appcompat.app.AppCompatActivity
import constantin.uvcintegration.TranscodeService
import constantin.video.core.AVideoSettings
import constantin.video.core.IsConnected
import constantin.video.core.player.VideoSettings
import constantin.video.example.databinding.ActivityMainBinding
import constantin.video.example.decodingperf.VideoActivityWithDatabase
import constantin.video.example.decodingperf.ViewBenchmarkDataActivity
import constantin.video.transmitter.AVideoTransmitter
import constantin.video.transmitter.AVideoTransmitterSettings

class MainActivity : AppCompatActivity(), View.OnClickListener {
    private lateinit var binding: ActivityMainBinding

    private lateinit var context: Context;
    private val permissionHelper= constantin.helper.RequestPermissionHelper(
            arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE, Manifest.permission.READ_EXTERNAL_STORAGE,
                    Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.CAMERA)
    );

    @SuppressLint("ApplySharedPref")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        permissionHelper.checkAndRequestPermissions(this)
        VideoSettings.initializePreferences(this, false)
        context = this

        val adapter = ArrayAdapter.createFromResource(this,
                R.array.video_test_files, android.R.layout.simple_spinner_item)


        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        binding.sVideoFileSelector.adapter = adapter
        //

        binding.bStartVideoActivity.setOnClickListener {
            val selectedTestFile = binding.sVideoFileSelector.getSelectedItem() as String
            val preferences = getSharedPreferences("pref_video", Context.MODE_PRIVATE)
            preferences.edit().putString(getString(R.string.VS_ASSETS_FILENAME_TEST_ONLY), selectedTestFile).commit()
            val intentVideoActivity = Intent()
            intentVideoActivity.setClass(context, VideoActivityWithDatabase::class.java)
            startActivity(intentVideoActivity)
        }
        binding.bStartSettingsActivity.setOnClickListener {
            val intent = Intent()
            intent.setClass(context, AVideoSettings::class.java)
            intent.putExtra(AVideoSettings.EXTRA_KEY, true)
            startActivity(intent)
        }
        binding.bStartTelemetry.setOnClickListener { IsConnected.openUSBTetherSettings(context) }
        binding.bStartViewDatabase.setOnClickListener { startActivity(Intent().setClass(context, ViewBenchmarkDataActivity::class.java)) }
        binding.bStartColorFormatsTester.setOnClickListener { startActivity(Intent().setClass(context, AColorFormatTester::class.java)) }
        binding.bStartTelemetry.setOnClickListener { startActivity(Intent().setClass(context, ATelemetryExample::class.java)) }
        binding.bStartGraphView.setOnClickListener { startActivity(Intent().setClass(context, GraphViewA::class.java)) }

        binding.bStartTranscodeService.setOnClickListener {
            TranscodeService.startTranscoding(context, "");
        }
        binding.bStopTranscodeService.setOnClickListener {
            TranscodeService.stopTranscoding(context);
        }
        binding.bStartStreamingServer.setOnClickListener {
            val intent = Intent().setClass(context, AVideoTransmitter::class.java)
            startActivity(intent)
        }
        binding.bStreamingServerSettings.setOnClickListener {
            // User chose the "Settings" item, show the app settings UI...
            val i = Intent()
            i.setClass(this, AVideoTransmitterSettings::class.java)
            startActivity(i)
        }
        binding.bTestFEC.setOnClickListener {
            //TestFEC.nativeTestFec();
        }
    }

    override fun onResume() {
        binding.sVideoFileSelector.isEnabled =
                getSharedPreferences("pref_video", Context.MODE_PRIVATE).getInt(getString(R.string.VS_SOURCE), 0) == 2
        super.onResume()
    }

    override fun onClick(v: View) {}

    override fun onRequestPermissionsResult(requestCode: Int,
                                            permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        permissionHelper.onRequestPermissionsResult(requestCode, permissions, grantResults)
    }

    public companion object {
        fun VIDEO_TEST_FILES_FOR_DB(context: Context): Array<String> {
            return context.resources.getStringArray(R.array.video_test_files_for_db)
        }
    }
}
